### PR TITLE
testing shutdown timeout

### PR DESCRIPTION
**Pull request by urso**

### CODE REVIEW

File: filebeat/beater/filebeat.go [hunk 1]
```diff
@@ -126,9 +126,13 @@ func (fb *Filebeat) Run(b *beat.Beat) error {
 	<-fb.done
 
 	if fb.config.ShutdownTimeout > 0 {
+		logp.Info("Shutdown timeout %v", fb.config.ShutdownTimeout)
```
Comment (ruflin) [inline filebeat/beater/filebeat.go]: Shutdown timeout started: %v

File: filebeat/beater/filebeat.go [hunk 2]
```diff
@@ -126,9 +126,13 @@ func (fb *Filebeat) Run(b *beat.Beat) error {
...
+
 		// Wait for either timeout or all events having been ACKed by outputs.
-		fb.sigWait.Add(wgEvents.Wait)
-		fb.sigWait.AddTimeout(fb.config.ShutdownTimeout)
+		fb.sigWait.Add(withLog(wgEvents.Wait,
+			"Continue shutdown: All enqueued events being published."))
+		fb.sigWait.Add(withLog(waitDuration(fb.config.ShutdownTimeout),
+			"Continue shutdown: shutdown timeout."))
```
Comment (ruflin) [inline filebeat/beater/filebeat.go]: "Continue shutdown: Shutdown timeout reached before all events published."

File: filebeat/tests/system/test_shutdown.py [hunk 3]
```diff
@@ -22,18 +20,56 @@ def test_shutdown(self):
...
+            ignore_older="1h",
+            shutdown_timeout="10s",
+        )
+        proc = self.start_beat(logging_args=["-e", "-v"])
+        time.sleep(.5)
+        proc.check_kill_and_wait()
+
+        log = self.get_log()
+        assert "Shutdown timeout 10s" in log
+        assert "Continue shutdown: All enqueued events being published." in log
```
Comment (ruflin) [inline filebeat/tests/system/test_shutdown.py]: Could you also check the registry file that all events were presisted there (offset)?

File: filebeat/tests/system/test_shutdown.py [hunk 4]
```diff
@@ -22,18 +20,56 @@ def test_shutdown(self):
...
+            ignore_older="1h",
+            shutdown_timeout="1s",
+        )
+        proc = self.start_beat(logging_args=["-e", "-v"])
+        time.sleep(.5)
+        proc.check_kill_and_wait()
+
+        log = self.get_log()
+        assert "Shutdown timeout 1s" in log
+        assert "Continue shutdown: shutdown timeout." in log
```
Comment (ruflin) [inline filebeat/tests/system/test_shutdown.py]: Could you check that the registry does not contain all events yet (offset)?

**Comment by ruflin** [comment]
It looks like the jenkins test somehow managed to end the test with a panic.

File: filebeat/beater/filebeat.go (hunk 1 above)
Comment (urso) [inline filebeat/beater/filebeat.go]: when using `Shutdown timeout started` I would assume `%v` to be `time.Now()`, right?

File: filebeat/tests/system/test_shutdown.py (hunk 3 above)
Comment (urso) [inline filebeat/tests/system/test_shutdown.py]: any helpers for testing this?

File: filebeat/tests/system/test_shutdown.py (hunk 4 above)
Comment (urso) [inline filebeat/tests/system/test_shutdown.py]: any helpers for testing this?

File: filebeat/beater/filebeat.go (hunk 1 above)
Comment (ruflin) [inline filebeat/beater/filebeat.go]: `Shutdown timeout strated. Waiting %v`?

File: filebeat/tests/system/test_shutdown.py (hunk 4 above)
Comment (ruflin) [inline filebeat/tests/system/test_shutdown.py]: `self.get_registry()`: https://github.com/elastic/beats/blob/master/filebeat/tests/system/test_crawler.py#L326

**Comment by urso** [comment]
There has been a race on shutdown with crawlers not being stopped yet, but `wgEvents.Wait()` being  run with potentially no events currently scheduled. The effect might be a panic on shutdown.

File: filebeat/tests/system/config/filebeat.yml.j2 [hunk 5]
```diff
@@ -132,11 +133,16 @@ processors:
 # Multiple outputs may be used.
 
 #------------------------------- File output ----------------------------------
+{%- if logstash %}
```
Comment (ruflin) [inline filebeat/tests/system/config/filebeat.yml.j2]: not that we have enabled: true again, we probably don't need to full block comments ;-)

**Comment by ruflin** [comment]
LGTM

File: filebeat/beater/filebeat.go [hunk 6]
```diff
@@ -119,16 +119,21 @@ func (fb *Filebeat) Run(b *beat.Beat) error {
 	if err != nil {
 		return err
 	}
-	// Stop crawler -> stop prospectors -> stop harvesters
-	defer crawler.Stop()
-
 	// Blocks progressing. As soon as channel is closed, all defer statements come into play
+
+	// Stop crawler -> stop prospectors -> stop harvesters
```
Comment (ruflin) [inline filebeat/beater/filebeat.go]: that should go one line down

File: filebeat/beater/filebeat.go [hunk 7]
```diff
@@ -119,16 +119,21 @@ func (fb *Filebeat) Run(b *beat.Beat) error {
...
 	<-fb.done
+	crawler.Stop()
```
Comment (ruflin) [inline filebeat/beater/filebeat.go]: Can you please add a note here why we don't use defer for the crawler.Stop? I already see me adding a defer to it in 3 months :-)

File: filebeat/beater/filebeat.go (hunk 6 above)
Comment (urso) [inline filebeat/beater/filebeat.go]: done

File: filebeat/beater/filebeat.go (hunk 7 above)
Comment (urso) [inline filebeat/beater/filebeat.go]: done

File: filebeat/tests/system/config/filebeat.yml.j2 (hunk 5 above)
Comment (urso) [inline filebeat/tests/system/config/filebeat.yml.j2]: as template variable `logstash.host` is used to configure the hostname, I'd keep it like this.

**Pull request merged**